### PR TITLE
Recsys dcg

### PR DESCRIPTION
**Pull request by zkid18**
## Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md)?
- [x] Did you check the code style? `catalyst-make-codestyle && catalyst-check-codestyle` (`pip install -U catalyst-codestyle`).
- [x] Did you make sure to update the docs? We use Google format for all the methods and classes.
- [x] Did you check the docs with `make check-docs`?
- [x] Did you write any new necessary tests?
- [x] Did you check that your code passes the unit tests `pytest .` ? 
- [x] Did you add your new functionality to the docs?
- [ ] Did you update the [CHANGELOG](https://github.com/catalyst-team/catalyst/blob/master/CHANGELOG.md)?

## Type of Change

- [ ] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

PS
- [x] I know, that I could [join slack](https://join.slack.com/t/catalyst-team-core/shared_invite/zt-d9miirnn-z86oKDzFMKlMG4fgFdZafw) for pull request discussion.

### CODE REVIEW

File: catalyst/metrics/ndcg.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    * :func:`ndcg`
+"""
+from typing import Sequence
+
+import torch
+
+
+def dcg(
+    outputs: torch.Tensor, targets: torch.Tensor, 
+    gain_function=lambda x: torch.pow(2, x) - 1, k=100
```
Comment (Scitator) [inline catalyst/metrics/ndcg.py]: types please

Comment (Scitator) [reply catalyst/metrics/ndcg.py]: maybe `k=10`?

File: catalyst/metrics/ndcg.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+) -> torch.Tensor:
+    """
+    Computes DCG@topk for the specified values of `topk`.
```
Comment (Scitator) [inline catalyst/metrics/ndcg.py]: `k` and `topk` naming? :)

File: catalyst/metrics/ndcg.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+            callable, gain function for the ground truth labels. 
+            on the deafult, the torch.pow(2, x) - 1 function used
+            to get emphasise on the retirvng the revelant documnets.
+        k (int):
+            Parameter fro evaluation on top-k items
+
+    Returns:
+        list with computed dcg@topk
+    """
+    k = min(outputs.size(1), k)
```
Comment (Scitator) [inline catalyst/metrics/ndcg.py:44]: should we assert `outputs.size(1) > k`?

Comment (Scitator) [reply catalyst/metrics/ndcg.py:44]: or warning?

File: catalyst/metrics/ndcg.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    sum_dcg =  torch.sum(discounted_gains, dim=1)
+    return sum_dcg
+
+
+def ndcg(
+    outputs: torch.Tensor, targets: torch.Tensor, 
+    gain_function=lambda x: torch.pow(2, x) - 1, k=100
+) -> torch.Tensor:
+    """
+    Computes nDCG@topk for the specified values of `topk`.
```
Comment (Scitator) [inline catalyst/metrics/ndcg.py]: `topK` :)

File: catalyst/metrics/ndcg.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+            on the deafult, the torch.pow(2, x) - 1 function used
+            to get emphasise on the retirvng the revelant documnets.
+        k (int):
+            Parameter fro evaluation on top-k items
+
+    Returns:
+        list with computed ndcg@topk
+    """
+    ideal_dcgs = dcg(targets, targets, gain_function, k)
+    ndcg = dcg(outputs, targets, gain_function, k) / ideal_dcgs
```
Comment (Scitator) [inline catalyst/metrics/ndcg.py]: could we make something like
```
predicted_dcgs = dcg(outputs, targets, gain_function, k)
ndcg = predicted_dcgs / ideal_dcgs
```

File: catalyst/metrics/tests/test_ndcg.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
+import math
```
Comment (Scitator) [inline catalyst/metrics/tests/test_ndcg.py:1]: could we also add tests for dcg?

**Review by Scitator — DISMISSED** [review]

**Review by Scitator — DISMISSED** [review]

File: catalyst/metrics/ndcg.py (hunk 3 above)
Comment (zkid18) [reply catalyst/metrics/ndcg.py:44]: Will be fixed in PR for preprocessing pipeline

**Pull request merged**